File: crack-data-structures-and-algorithms/leetcode/next_permutation_q31.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        x, y = -1, -1
        # 1) find largest index x, s.t nums[x] < numx[x+1]
        # 2) find largest index y, s.t nums[x] < nums[y]; and because loop here terminates
        # before the last element, we need to check the last element for y particularly.
        for i in range(len(nums) - 1):
            if nums[i] < nums[i+1]:
                x = i

            if x != -1 and nums[i] > nums[x]:
                y = i

        if x == -1:
            nums.reverse()
            return

        # Because for loop only checks [0, |nums|-2], which skips the last element
        # we need to amend it manually.
        if nums[-1] > nums[x]:
            y = len(nums) - 1

        nums[x], nums[y] = nums[y], nums[x]
        reverse_part(nums, x+1, len(nums)-1)


def reverse_part(nums, l, r):
    while l < r:
        nums[l], nums[r] = nums[r], nums[l]
        l += 1
        r -= 1

File: crack-data-structures-and-algorithms/leetcode/next_permutation_q31.cpp
#include <vector>

using namespace std;

// 核心思路
// 标准的 next_permutation 的算法
// 1) 找到一个最大的i，满足 a[i] < a[i+1]
//    如果这一步找不到说明当前序列已经是最后一个 permutation，需要对数组做一次 reverse 转换为第一个
// 2) 从 i+1 开始找到一个最大的 j，满足 a[i] < a[j]；因为(1)满足所以一定能找到这个j
// 3) a[i] <-> a[j]，然后 reverse a[i+1...n)
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int base = -1;
        for (int i = 0; i < nums.size() - 1; ++i) {
            if (nums[i] < nums[i+1]) {
                base = i;
            }
        }

        if (base == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }

        int target = base;
        for (int j = base + 1; j < nums.size(); ++j) {
            if (nums[base] < nums[j]) {
                target = j;
            }
        }

        swap(nums[base], nums[target]);
        reverse(&nums[base+1], &nums.back() + 1);

        return;
    }
};
